docs/security/index.md

```diff
@@ -0,0 +1,453 @@
+retrieve user attributes to provide label enforcement decisions.
+
+Therefore as part of this new API we adopt a similar approach to [Labels](#label-format--syntax) in that we move to
+treat user entitlements as opaque byte sequences with responsibility for retrieving and interpreting user entitlements
+left to plugins via their `EntitlementsProvider` and `EntitlementsParser` implementations.  This means that the burden
+of configuring any necessary supporting services, e.g. connectivity to Access, becomes an internal implementation detail
+of a plugin, rather than an application concern. Entitlements use the same optional schema prefix as part of the byte
+sequences to allow plugins to distinguish between different entitlement schemas they may support, and this allows for
+future plugins to potentially combine multiple entitlement providers.
+
+As with labels the `Entitlements<T>` interface treats entitlements in abstract terms from an application perspective
+while allowing a plugin implementation to decode the entitlements into whatever data structure(s) are appropriate. A
+plugins implementation of the `EntitlementsParser` interface can supply whatever decoding logic it needs for its
+implementation.
+
+### Data Access Enforcement
+
+Data Access Enforcement is done by combining the users retrieved [Entitlements](#entitlements) with the
+[Labels](#label-format--syntax) present on the data.  This is done by first preparing an `Authorizer` via the
+`prepareAuthorizer()` method passing in the users entitlements.  `Authorizer` instances are intended to be scoped to the
+lifetime of a single user request so they may cache any access decisions if they encounter the same labels repeatedly.
+
+Once an application has an `Authorizer` it calls the `canAccess()` method passing in the labels for the data it needs an
+access decision for.  This returns either `true` for accessible, or `false` for forbidden.  In the event of any
+problem/ambiguity in computing an access decisions plugins **MUST** fail-safe by defaulting to returning `false` if they
+can't make an access decision.
+
+You can see the complete lifecycle outlined later in [Example Usage](#example-usage).
+
+### Plugin Loading
+
+Plugins are loaded by applications by calling `SecurityPluginLoader.load()` which uses the JVM `ServiceLoader` to locate
+registered plugins.  The loaded plugin is cached for the lifetime of the JVM so subsequent calls just return the
+previously loaded instance.
+
+The expectation is that there is `1`, and **ONLY** `1` plugin registered in this way.  If no plugin is registered, more
+than 1 plugin is registered, or loading the registered plugin throws an error, then the system defaults to a fail-safe
+mode and instead loads the unregistered `FailSafePlugin`. In fail-safe mode the system treats all labels and
+entitlements as invalid and defaults all access decisions to forbidden.  This will be clearly and explicitly noted in
```

Comment:
> Given how vital this is should we force the app to shutdown if there is an issue?

Throwing an error is our best effort to do this.  We could `System.exit()` but that tends to make writing tests around the behaviour very difficult.

While we control our applications customers can build their own applications that integrate our security model using this API and we won't be able to control how they call the API.  So falling back into a fail-safe mode is a compromise solution to try and protect developers from themselves.

Reply:
> What happens if we can't connect to Kafka or Access? I guess we should be consistent

For Kafka it depends on the application, generally it will just keep trying to connect and many eventually time out and crash restart.

For Access the application tends to get an error retrieving user attributes, again depending on the application this either manifests as a user visible error or their request is treated as having no attributes so they get minimal access to data.